draft WIP feat: Preemptively filter for pushdown-preventing columns in ListingTable

itsjunetime opened the pull request:
Just making this PR to facilitate discussion before actually trying to make a PR against apache's repo

itsjunetime committed:
feat: Preemptively filter for pushdown-preventing columns in ListingTable

alamb committed:
Minor: update sqllogictest to treat Utf8View as text (#12033)

* Minor: update sqllogictest to treat Utf8View as text

* fmt

Commit:
[MINOR] correct document mistakes (#12068)

* mistake

* mistake

alamb commented on datafusion/core/src/datasource/listing/helpers.rs:
```diff
@@ -51,12 +52,26 @@ use object_store::{ObjectMeta, ObjectStore};
 /// - the expression can be marked as `TableProviderFilterPushDown::Exact` once this filtering
 ///   was performed
-pub fn expr_applicable_for_cols(col_names: &[String], expr: &Expr) -> bool {
+pub fn expr_applicable_for_schemas(
```
I think one thing that is confusing here is that [`ListingTable::supports_filters_pushdown`](https://github.com/alamb/datafusion/blob/6877e574fc60b0845fc538d40ce371552cc0620a/datafusion/core/src/datasource/listing/table.rs#L832-L838) is returning `TableProviderFilterPushDown::Exact` if the predicate can be applied to "partitioning" columns

In this case, "partition pruning" means if the data was stored in [multiple files as described](https://github.com/apache/datafusion/blob/6877e574fc60b0845fc538d40ce371552cc0620a/datafusion/core/src/datasource/listing/table.rs#L544-L551) here

```
/// table2
///  ├── date=2024-06-01
///  │    ├── file3.parquet
///  │    └── file4.parquet
///  └── date=2024-06-02
///       └── file5.parquet
```

So a predicate such as `date = '2024-06-01'` could be exact as it could filter directories exactly

alamb replied:
It seems like you are trying to extend the meaning of `expr_applicable_for_schemas` to also include these push down predicates in Parquet, but that combines two different pruning concepts

Another potential way to plumb this in, might be easier is to leave the existing partition pruning checks and instead extend [`ListingTable::supports_filters_pushdown` ](https://github.com/apache/datafusion/blob/6877e574fc60b0845fc538d40ce371552cc0620a/datafusion/core/src/datasource/listing/table.rs#L832-L842) to delegate to the `FileFormat`:
 

```rust
 fn supports_filters_pushdown(
        &self,
        filters: &[&Expr],
    ) -> Result<Vec<TableProviderFilterPushDown>> {
        Ok(filters
            .iter()
            .map(|filter| {
                if expr_applicable_for_schemas(
                    filter,
                    &self.file_schema,
                    &self.table_schema,
                ) {
                    // if filter can be handled by partition pruning, it is exact
                    TableProviderFilterPushDown::Exact
                } else {
                    // ** NOTE THIS IS CHANGED ****
                    // Some formats like Parquet can exactly apply filters
                    // when row level filtering is enabled
                    self.format.supports_filters_pushdown(filter)
                }
            })
            .collect())
    }
```

Then add a function in `FileFormat` trait for supporting pushdown that by default returns `TableProviderFilterPushDown::InExact`

alamb commented on datafusion/core/src/datasource/physical_plan/parquet/row_filter.rs:
```diff
@@ -268,90 +260,156 @@ impl<'a> FilterCandidateBuilder<'a> {
+    }
+}
+
+// a struct that implements TreeNodeRewriter to traverse a PhysicalExpr tree structure to determine
```
Nice

alamb commented on datafusion/core/src/datasource/physical_plan/parquet/row_filter.rs:
```diff
@@ -268,90 +260,156 @@ impl<'a> FilterCandidateBuilder<'a> {
-                can_use_index,
-                projection: self.required_column_indices.into_iter().collect(),
-            }))
+    pub fn build(self, metadata: &ParquetMetaData) -> Result<Option<FilterCandidate>> {
```
One thing that may be necessary is being able to check for pushdown"ability" without having the parquet metadata if possible

aalexandrov committed:
Plan `LATERAL` subqueries (#11456)

* Planner: support `LATERAL` subqueries

* Planner: use `DFSchema::merge` in `create_relation_subquery`

In order to compute the `set_outer_from_schema` argument we currently
use `DFSchema::join`. When we combine the current outer FROM schema with
the current outer query schema columns from the latter should override
columns from the first, so the correct way is to use `DFSchema::merge`.

To witness the fix, note that the query in the fixed test case isn't
planned as expected without the accompanying changes.

* Update plans

---------

Co-authored-by: Andrew Lamb <andrew@nerdnetworks.org>

2010YOUY01 committed:
vectorize random() scalar function (#12078)

nrc committed:
functions: support strpos with mixed string types (#12072)

Signed-off-by: Nick Cameron <nrc@ncameron.org>

findepi committed:
Update to `clap` 4.5.16 (#12064)

itsjunetime committed:
Fix behavior to make all previous tests work and lay groundwork for future tests
